src/sparse/csmat.rs

```diff
@@ -1139,6 +1139,30 @@ where
         })
     }
 
+    /// Get the diagonal of a sparse matrix
+    pub fn diag(&self) -> CsVecI<N, I>
+    where
+        N: Clone,
+    {
+        let shape = self.shape();
+        let smallest_dim: usize = cmp::min(shape.0, shape.1);
+        let mut index_vec = Vec::new();
+        let mut data_vec = Vec::new();
```

Comment:
I think there could be a heuristic to create `index_vec` and `data_vec` with `Vec::with_capacity`, to minimize reallocations? It's frequent to have sparse matrices with lots of elements on the diagonal. So maybe `let capacity = smallest_dim / 2;` ? What do you think?

Reply:
That is a great idea. I am even tempted to preallocate the max possible size, `smallest_dim`.  I presume this would be the most performant, and least memory concerned? I'll defer to you to decide as I'm not an expert.

Reply:
It's hard to guess without enough representative tests. I think this is good for now, and if in the future there comes a need to customize this heuristic we can always add another function with an additional parameter. It's nice that you added `shrink_to_fit`, this will probably be important when there are not that many diagonal elements.